controlplane/kubeadm/internal/cluster.go

```diff
@@ -450,6 +430,78 @@ func (c *cluster) getEtcdClientForNode(nodeName string, tlsConfig *tls.Config) (
 	return customClient, nil
 }
 
+// updateKubeadmConfigMap gets, mutates, and updates the cluster's kubeadm
+// ConfigMap.
+// Copied from https://github.com/vmware/cluster-api-upgrade-tool/pkg/upgrade/control_plane.go
+func (c *cluster) updateKubeadmConfigMap(ctx context.Context, f func(in *corev1.ConfigMap) (*corev1.ConfigMap, error)) error {
```

Comment:
An alternative to using a closure here would be to extract the configmap operations into their own type. 

The resulting layers become: 
 * The `cluster` object is responsible for interacting with the apiserver (Read, Update)
 * A new struct has methods for modifications to a kubeadmconfigmap. (New is basically the `extractKubeadmConfigMapClusterStatus`, then `removeNodeFromKubeadmConfigMapClusterStatusAPIEndpoints` is a method on the type.

This keeps the layers cleanly separated. It's very similar to what you've got here but instead of a generic update function we define the specific actions we need without abstracting too early.

If the above was not clear, this is the pseudocode of what was described.
```
configMap := client.Get(configmap, ... )
kcfg := NewKubeadmConfig(configMap)
kcfg.RemoveAPIEndpoint(machine.status.noderef.name)
client.Update(kcfg)
```

Some of the benefits of this approach are that the functions acting on the kubeadm config are attached to a struct which implies context. Thus, we can avoid long names like `removeNodeFromKubeadmConfigMapClusterStatusAPIEndpoints`. The layers between each component are clean and the kubeadmconfig object need not know anything except what a kubeadmconfig is and how it's structured. It can be easily tested in isolation. The code becomes imperative without relying on descriptive function names.

Reply:
Created #2429 to track this as a followup